node/src/main/scala/org/bitcoins/node/networking/peer/MerkleBuffers.scala

```diff
@@ -0,0 +1,111 @@
+  * corresponding transactions are sent immediately after. That means
+  * we have to correlate the received merkleblocks with the matching
+  * transactions.
+  *
+  * This buffer is responsible for calling the approriate callbacks
+  * once a merkle block has received all its transactions.
+  */
+private[peer] object MerkleBuffers extends BitcoinSLogger {
+  private type MerkleBlocksWithTransactions =
+    mutable.Map[MerkleBlock, mutable.Builder[Transaction, Vector[Transaction]]]
+
+  private val underlyingMap: MerkleBlocksWithTransactions = mutable.Map.empty
+
+  /** Adds the given merkleblock to the buffer */
+  def putMerkle(merkle: MerkleBlock): Unit = {
+    val tree = merkle.partialMerkleTree
+    val matches = tree.extractMatches
+
+    logger.debug(s"Received merkle block, expecting ${matches.length} TX(s)")
+
+    if (matches.nonEmpty) {
+      logger.trace(s"Adding merkleBlock=${merkle.blockHeader.hashBE} to buffer")
+      underlyingMap.put(merkle, Vector.newBuilder)
+    } else {
+      logger.trace(
+        s"Merkleblock=${merkle.blockHeader.hashBE} has no matches, not adding to buffer")
+    }
+
+    ()
+  }
+
+  /** Attempts to add the given transaction to a corresponding
+    * merkleblock in the buffer.
+    *
+    * @param tx The transaction to (maybe) add to the buffer
+    * @param callbacks The callbacks to execute if we're
+    *        finished processing a merkleblock
+    *
+    * @return If the transaction matches a merkle block, returns true.
+    *         Otherwise, false.
+    */
+  def putTx(
+      tx: Transaction,
+      callbacks: Seq[DataMessageHandler.OnMerkleBlockReceived]): Boolean = {
+    val blocksInBuffer = underlyingMap.keys.toList
+    logger.trace(s"Looking for transaction=${tx.txIdBE} in merkleblock buffer")
+    logger.trace(s"Merkleblocks in buffer: ${blocksInBuffer.length}")
+    blocksInBuffer.find { block =>
+      val matches = block.partialMerkleTree.extractMatches
+
+      logger.trace(
+        s"Block=${block.blockHeader.hashBE} has matches=${matches.map(_.flip)}")
+
+      matches.exists(_ == tx.txId)
+    } match {
+      case None =>
+        logger.debug(
+          s"Transaction=${tx.txIdBE} does not belong to any merkle block")
+        false
+      case Some(key) =>
+        handleMerkleMatch(tx, merkleBlock = key, callbacks = callbacks)
+    }
+  }
+
+  // TODO Scaladoc
+  private def handleMerkleMatch(
+      transaction: Transaction,
+      merkleBlock: MerkleBlock,
+      callbacks: Seq[DataMessageHandler.OnMerkleBlockReceived]) = {
+    val merkleBlockMatches = merkleBlock.partialMerkleTree.extractMatches
+    val merkleHash = merkleBlock.blockHeader.hashBE
+
+    val txHash = transaction.txIdBE
+
+    logger.debug(s"Transaction=$txHash matched merkleBlock=$merkleHash")
+
+    logger.trace(s"Adding transaction=$txHash to buffer")
+    val builder = underlyingMap(merkleBlock) // TODO: error handling
+    builder += transaction
+
+    val transactionSoFar = builder.result()
```

Comment:
do we want to call `builder.result()` here? If there are multiple matches this will force the vector builder to be cleared i believe 

```
  /** Produces a collection from the added elements.  This is a terminal operation:
   *  the builder's contents are undefined after this operation, and no further
   *  methods should be called.
   *  
   *  @return a collection containing the elements added to this builder.
   */
```

Reply:
```scala
@ val builder = Vector.newBuilder[Int] 
builder: collection.mutable.Builder[Int, Vector[Int]] = scala.collection.immutable.VectorBuilder@a240452

@ builder += 1 
res1: collection.mutable.Builder[Int, Vector[Int]] = scala.collection.immutable.VectorBuilder@a240452

@ builder += 2 
res2: collection.mutable.Builder[Int, Vector[Int]] = scala.collection.immutable.VectorBuilder@a240452

@ builder += 3 
res3: collection.mutable.Builder[Int, Vector[Int]] = scala.collection.immutable.VectorBuilder@a240452

@ builder.result() 
res4: Vector[Int] = Vector(1, 2, 3)

@ builder.result() 
res5: Vector[Int] = Vector(1, 2, 3)

@ builder.result() 
res6: Vector[Int] = Vector(1, 2, 3)
```

Although you do have a point. Could replace with a mutable buffer instead of a builder?

Reply:
Actually it looks like VectorBuilder is safe, we just need to type the map as `mutable.ReusableBuilder` rather than `mutable.Builder`